pkg/synchronizer/opstate.go

```diff
@@ -0,0 +1,198 @@
+	if device == nil {
+		// Can happen if the device is reported in prometheus, but does not exist in
+		// ROC. Rare in real life, but plausible in test infrastructure.
+		// Message is currently disabled, as it is fairly noisy.
+		//log.Warnf("Failed to find device %v in site %s of enterprise %s", imsi, siteID, entID)
+		return
+	}
+
+	if device.State == nil {
+		device.State = &DeviceState{}
+	}
+	device.State.IpAddress = aStr(mobileIP)
+
+	if connected != 0 {
+		device.State.Connected = aStr("Yes")
+	} else {
+		device.State.Connected = aStr("No")
+	}
+}
+
+// poll promtheus for metrics, and add them to the config tree
+func (s *Synchronizer) pollPrometheus(config *gnmi.ConfigForest, prom *prometheusInfo) error {
+	log.Debugf("Polling prometheus site %s enterprise %s endpoint %s", prom.siteID, prom.entID, prom.endpoint)
+
+	fetcher, err := s.getFetcher(prom.endpoint)
+	if err != nil {
+		return err
+	}
+	v, err := fetcher.GetVector("subscribers_info")
+	if err != nil {
+		return err
+	}
+
+	// vector is []*Sample
+	// sample is {Metric Value:SampleValue Time}
+	// metric is a labelset
+	// labelsite is map[labelname]labelvalue
+	// samplevalue is float64
+
+	for _, sample := range v {
+		// labels will be imsi and mobile_ip
+		imsiString, okay := sample.Metric["imsi"]
+		if !okay {
+			log.Warn("Failed to get imsi from sample")
+			continue
+		}
+		mobileIP, okay := sample.Metric["mobile_ip"]
+		if !okay {
+			log.Warn("Failed to get mobile_ip from sample")
+			continue
+		}
+
+		imsi, err := strconv.ParseUint(string(imsiString), 10, 64)
+		if err != nil {
+			log.Warn("Failed to convert imsi from sample")
+			continue
+		}
+
+		s.updateDeviceIPAddress(config, prom.entID, prom.siteID, imsi, string(mobileIP), float64(sample.Value))
+	}
+
+	return nil
+}
+
+// build a list of all prometheus instances that we want to poll
+func (s *Synchronizer) buildPrometheusList(config *gnmi.ConfigForest) []*prometheusInfo {
+	config.Mu.Lock()
+	defer config.Mu.Unlock()
+
+	prometheusList := []*prometheusInfo{}
+
+	for entID, entRoot := range config.Configs {
+		enterprise := entRoot.(*RootDevice)
+		for siteID, site := range enterprise.Site {
+			if site.ConnectivityService != nil &&
+				site.ConnectivityService.Core_4G != nil &&
+				site.ConnectivityService.Core_4G.AccPrometheusUrl != nil &&
+				*site.ConnectivityService.Core_4G.AccPrometheusUrl != "" {
+				prometheusList = append(prometheusList, &prometheusInfo{entID: entID, siteID: siteID, endpoint: *site.ConnectivityService.Core_4G.AccPrometheusUrl})
+			}
+		}
+	}
+
+	return prometheusList
+}
+
+func (s *Synchronizer) pollAllPrometheus(config *gnmi.ConfigForest) {
+	promList := s.buildPrometheusList(config)
+
+	// Walk the list of prometheus instances. We do this without holding the
+	// lock, so we do not block the config trees while waiting on prometheus.
+	// We will lock the tree when we go to update a value.
+	for _, promInfo := range promList {
+		err := s.pollPrometheus(config, promInfo)
```

Comment:
I wonder if it would be better for each call to have its own go routine.
The sequential nature of this loop is not really needed

Reply:
That's a good suggestion. My gut feeling is that this implementation is going to be rather short-lived (soon to be replaced by Kafka and/or analytics engine), and generally used in deployments where there is only one ACC. The challenge is going to be what to do if there are multiple sites that all use the same ACC. The behavior will be correct, but it will also be wasteful. Let me think on this.

Reply:
OK